shared/video-sdk/develop/stream-raw-audio-and-video/project-implementation/windows.mdx

```diff
@@ -326,49 +274,45 @@ To register and use video and audio frame observers in your <Vpl k="CLIENT" />,
             mediaEngine.queryInterface(agoraEngine, agora::rtc::AGORA_IID_MEDIA_ENGINE);
             int nRet = 0;
             agora::base::AParameter apm(agoraEngine);
-            if (mediaEngine.get() == NULL)
-                return FALSE;
-            if (bEnable) {
-                
+            if (mediaEngine.get() == NULL) return FALSE;
+            if (bEnable) 
+            {
                 // Register video & audio frame observer
-                nRet = mediaEngine->registerVideoFrameObserver(video_frame_observer);
-                nRet = mediaEngine->registerAudioFrameObserver(audio_frame_observer);
-
-                // Set the format of the captured raw audio data.
-                int SAMPLE_RATE = 16000, SAMPLE_NUM_OF_CHANNEL = 1, SAMPLES_PER_CALL = 1024;
+                nRet = mediaEngine->registerVideoFrameObserver(new CustomVideoFrameObserver());
+                nRet = mediaEngine->registerAudioFrameObserver(new CustomAudioFrameObserver());
 
-                agoraEngine->setRecordingAudioFrameParameters(SAMPLE_RATE, SAMPLE_NUM_OF_CHANNEL,
-                    RAW_AUDIO_FRAME_OP_MODE_READ_WRITE, SAMPLES_PER_CALL);
+		        // Set the format of the captured raw audio data.
+		        int SAMPLE_RATE = 16000, SAMPLE_NUM_OF_CHANNEL = 1, SAMPLES_PER_CALL = 1024;
 
-                agoraEngine->setPlaybackAudioFrameParameters(SAMPLE_RATE, SAMPLE_NUM_OF_CHANNEL,
-                    RAW_AUDIO_FRAME_OP_MODE_READ_WRITE, SAMPLES_PER_CALL);
+		        agoraEngine->setRecordingAudioFrameParameters(SAMPLE_RATE, SAMPLE_NUM_OF_CHANNEL,
+                RAW_AUDIO_FRAME_OP_MODE_READ_WRITE, SAMPLES_PER_CALL);
 
-                agoraEngine->setMixedAudioFrameParameters(SAMPLE_RATE, SAMPLE_NUM_OF_CHANNEL, SAMPLES_PER_CALL);
+		        agoraEngine->setPlaybackAudioFrameParameters(SAMPLE_RATE, SAMPLE_NUM_OF_CHANNEL,
+                RAW_AUDIO_FRAME_OP_MODE_READ_WRITE, SAMPLES_PER_CALL);
 
+		        agoraEngine->setMixedAudioFrameParameters(SAMPLE_RATE, SAMPLE_NUM_OF_CHANNEL, SAMPLES_PER_CALL);
             }
-            else {
+            else 
+            {
                 nRet = mediaEngine->registerVideoFrameObserver(NULL);
                 nRet = mediaEngine->registerAudioFrameObserver(NULL);
             }
             return nRet == 0 ? TRUE : FALSE;
         }
         ```
 
-    1. To execute this method before joining the channel, add the following to the `Join` button event listener,
-  `OnBnClickedButton2()`, before `agoraEngine->enableAudio()`:
+    1. To register the observers at startup, in `AgoraImplementationDlg.cpp`, add the following at the of `setupVideoSDKEngine()`:
```

Comment:
Actually wrong at the very first removal fro original place. I kept this call in the `JOIN` button listener event. We can not keep it in `setupVideoSDKEngine()` or at the startup (`OnInitDialog()`). Because a user can join or leave the channel multiple time but OninitDlg will be called only once at the start of the system .Suppose a user left the channel and re-join again... then if it kept in start up , it will not enabled . For the same reason we should keep the disabling the observers `EnableAudioVideoCapture(FALSE)' only in the LEAVE` button listen event. I had kept them at correct place. Please don't make these changes.